kong/runloop/handler.lua

```diff
@@ -635,11 +639,24 @@ do
       end
     end
 
+    local counter = 0
+    local page_size = db.routes.pagination.page_size
     for route, err in db.routes:each() do
       if err then
         return nil, "could not load routes: " .. err
       end
 
+      if kong.core_cache and counter % page_size == 0 then
+        local new_version, err = get_router_version()
```

Comment:
I think it would be sufficient to only check this once after the loop, right? (before `Router.new`)

(In fact, I wonder if it wouldn't be more correct, even: wouldn't it be possible for the problem to happen even if the number of routers is less than one page size, but a change happens during that single yield, after the DB data has returned but before the scheduler resumed this coroutine?)

Reply:
The check is here, after each page, so that we will quit as soon as possible (to avoid doing unneccessary work, and release the request / buid). This code will `yield` on each `:page`, and not yield anywhere else. If we get last page (but the changes where made), it is ok, as we got full page (non-invalid). Similar to single page thing, if we got results from db, that result is ok, even if there were changes. The next request/build will start again.

Reply:
Ah, ok, got it! So the problem is really that in the case of concurrent modification, the `each` iteration as a whole returns an invalid set, correct?

Reply:
Yes, we cannot trust routes/plugins if state in those tables changes beween `:page` calls.

1. if there is a single page we can always trust (i will add a condition for that)
2. if things change when looping last page it doesn't matter

So yes `as a whole`.